PLAT-10367: Message Service implementation.

symphony-hong opened the pull request:
### Ticket
[PLAT-10367](https://perzoinc.atlassian.net/browse/PLAT-10367)

### Description
Message Service implementation

### Dependencies
List the other pull requests that should be merged before/along this one.

### Checklist
- [x] Referenced a ticket in the PR title and in the corresponding section
- [x] Filled properly the description and dependencies, if any
- [x] Unit tests updated or added
- [x] Docstrings added or updated
- [ ] Updated the documentation in [docs folder](../docs)

symphony-hong committed:
Message Service implementation.

Create MultiAttachmentsMessageApi class inherited MessagesApi allow bot to call the api which will send a message with multiple attachments and files as a list of files.

Create MessageService class wrapping all the service api related to message service which are documented in the https://developers.symphony.com/restapi/reference#messages-v4

Create method in the SymphonyBdk class to initialize the MessageService instance and provide it to bot developer to work with message service.

Create an example which shows how a bot interact with the message service.

Fix wrong path for agent api client.

symphony-youness commented on symphony/bdk/core/symphony_bdk.py:
```diff
@@ -26,6 +37,20 @@ async def bot_session(self) -> AuthSession:
+                PodApi(self._pod_client),
+                AttachmentsApi(self._agent_client),
+                DefaultApi(self._pod_client),
+                await self.bot_session()
```
Regarding the lazy init, I discussed it quickly with @symphony-elias and this works fine for now !!

symphony-youness commented on symphony/bdk/core/service/message_service.py:
```diff
@@ -0,0 +1,366 @@
+
+
+class MessageService:
+    def __init__(self, messages_api: MultiAttachmentsMessagesApi,
```
Just to check. I am having some confusion. 
What problem does the `MultiAttachmentsMessagesApi` solve that the generated `MessagesApi` has? I can't seem to find it on the java BDK.
Is it related to some new specs?

symphony-hong replied:
The generated `MessagesApi` doesn't allow to send a message with multiple attachments which is allowed by the api provided by the agent. That's why I create the `MultiAttachmentsMessagesApi` which accepts the attachment field as a list of file.

symphony-youri commented on symphony/bdk/core/service/message_service.py:
```diff
@@ -0,0 +1,366 @@
+        }
+        return await self._messages_api.v1_message_id_get(**params)
+
+    async def list_attachments(self, stream_id: str, **kwargs) -> [StreamAttachmentItem]:
```
is the kwargs temporary? if we use it don't we loose completion/documentation?

symphony-hong replied:
I don't really understand your point that the kwargs temporary. IMHO, the **kwargs contains all the optional fields that we gonna pass to the api and apparently we have to document all the fields that we can pass to the api through kwargs

symphony-youri replied:
and what about using default arguments? (named ones like here https://www.geeksforgeeks.org/default-arguments-in-python/)
could that help to be more explicit?

symphony-youri commented on symphony/bdk/core/service/message_service.py:
```diff
@@ -0,0 +1,366 @@
+from symphony.bdk.gen.pod_model.stream_attachment_item import StreamAttachmentItem
+
+
+class MultiAttachmentsMessagesApi(MessagesApi):
```
is it ok to put both classes in the same file? to keep it simple would not it be simpler to have a separate file?

symphony-hong replied:
The file containing the MessagesApi is generated from the specs and we shouldn't modify it, that's why I keep this class here.

symphony-youri replied:
and we cannot move it to its own file?

symphony-youri commented on symphony/bdk/core/service/message_service.py:
```diff
@@ -0,0 +1,366 @@
+                'http_method': 'POST',
+                'servers': None,
+            },
+            params_map={
```
looks like from here it is similar to above, maybe we can factorize it?

symphony-hong replied:
actually we have a slight difference between `sid` and `sids`

symphony-hong committed:
Move MultiAttachmentsMessagesApi class to dedicated python file

Move MultiAttachmentsMessagesApi class to symphony/bdk/core/service/message/multi_attachments_messages_api.py

Using default value parameters instead of kwargs

symphony-elias commented on symphony/bdk/core/service/message/message_service.py:
```diff
@@ -0,0 +1,217 @@
+            'version': version
+        }
+        if attachment is not None:
+            params['attachment'] = attachment
```
I'd prefer to have everything related to `attachment` in a single place like: remove line 70 and have:
```
if attachment is not None:
    params['attachment'] = attachment if type(attachment) is list else [attachment]
```
Same for `preview`.

symphony-elias commented on symphony/bdk/core/service/message/multi_attachments_messages_api.py:
```diff
@@ -0,0 +1,214 @@
+        super().__init__(api_client)
+
+        def __v4_message_create_post(self, **kwargs):
+            kwargs['async_req'] = kwargs.get(
```
Could you explain why this is needed? I was expecting the default values to be what you put here.

symphony-hong replied:
Actually I took inspiration from what was generated in the `MessagesApi` class, I refactored a bit but I decided to keep the fields that I'm not sure to be safe: https://github.com/SymphonyPlatformSolutions/symphony-api-client-python/blob/2.0/symphony/bdk/gen/agent_api/messages_api.py#L859-L877

symphony-elias replied:
👌 for me

symphony-elias commented on symphony/bdk/core/service/message/multi_attachments_messages_api.py:
```diff
@@ -0,0 +1,214 @@
+            kwargs['_check_return_type'] = kwargs.get(
+                '_check_return_type', True
+            )
+            kwargs['_host_index'] = kwargs.get('_host_index')
```
Please mind that it only puts a None value for key `_host_index` if it key was not existing before. Is this needed?

symphony-hong replied:
same answer as the comment above

symphony-elias replied:
👌

symphony-hong committed:
Unittest added

symphony-hong committed:
Documentation added

symphony-elias reviewed (approved):
Looks good !

symphony-hong committed:
Update markdown documentation

symphony-youri commented on symphony/bdk/core/symphony_bdk.py:
```diff
@@ -26,6 +37,26 @@ async def bot_session(self) -> AuthSession:
             self._bot_session = await self._authenticator_factory.get_bot_authenticator().authenticate_bot()
         return self._bot_session
 
+    async def messages(self):
```
is it ok to make it async?

For instance should we perform any remote call when creating the message service?
Or should we do the first calls only when actually using the service for the first time?

symphony-hong replied:
the problem here is we need a bot session to initialise the message service instance. So I guess, we should pass the `authenticator` to the message service constructor then inside the message service, each time we make a api call we gonna check the bot session there?

symphony-hong replied:
I think it's okay to leave this method async. In fact, it's only really async at the first time we authenticate the bot to get the `bot_session`.
We can also init the bot session as `AuthSession(bot_authenticator)`, but then inside the message, at the beginning of each service method, we have to check if the token is None or not, then authenticate the bot, and TBH I don't really like this solution.
WDYT?

symphony-youri replied:
What if the check is part of the getter in AuthSession?
One problem I see is that in case of concurrent access we might authenticate twice (if we don't lock) but with async it is not going to happen anyway

symphony-youri commented on symphony/bdk/core/symphony_bdk.py:
```diff
@@ -26,6 +37,26 @@ async def bot_session(self) -> AuthSession:
             self._bot_session = await self._authenticator_factory.get_bot_authenticator().authenticate_bot()
         return self._bot_session
 
+    async def messages(self):
```
Should we provide the returned type here?

symphony-youri replied:
(with the -> in the signature)

symphony-youri commented on docs/message_service.md:
```diff
@@ -0,0 +1,47 @@
+                )
+
+        finally:
+            await bdk.close_clients()
```
should we use a with: block (or a sort of automatic close)? (to avoid forgetting about closing the clients)

symphony-hong committed:
Using async context manager to initialise the BDK entry point

symphony-elias commented on symphony/bdk/core/service/message/message_service.py:
```diff
@@ -0,0 +1,340 @@
+            'skip': skip,
+            'limit': limit
+        }
+        message_list = await self._messages_api.v4_stream_sid_message_get(**params)
```
What about `self._messages_api.v4_stream_sid_message_get(sid=stream_id, since=since, session_token=self._auth_session.session_token, ...)`
IMHO having a param dict which is then unpacked makes things more complicated.

symphony-hong replied:
I would leave it as a param dict because for instance, `messages_api.v4_stream_sid_message_get` is defined as an instance of `Endpoint` callable class accepting the **kwargs parameters [Endpoint](https://github.com/SymphonyPlatformSolutions/symphony-api-client-python/blob/2.0/symphony/bdk/gen/api_client.py#L751-L762)
And IMHO, it doesn't impact much to the complexity of the code

symphony-elias replied:
I'd like to have someone else's opinion on that one, like @symphony-youri

symphony-youri replied:
no strong opinion on that because it is internal and not exposed to users (and I have tried to quickly search for guidelines on how to be idiomatic on this and did not find anything to help me decide)

We have other places where add parameters conditionally (for attachments) so maybe we still need a map?
I would perhaps have a small method to reuse to avoid repeating the session_token part everywhere?

symphony-hong committed:
Expose the close_client method for user to close the bdk manually

symphony-hong committed:
Lazy authentication bot

symphony-youri commented on symphony/bdk/core/auth/auth_session.py:
```diff
@@ -3,12 +3,32 @@ class AuthSession:
+        self._key_manager_token = await self._authenticator.retrieve_key_manager_token()
+
+    @property
+    async def session_token(self):
```
and now when I see as async I'm wondering if it is the right way to go (I'm lacking experience on async in Python here :)) 
let's go this way for now and we can always reevaluate that later
